chat: bring own database guide (#2949)
ably/docs: src/pages/docs/guides/chat/export-chat.mdx

## Decoding and storing messages

Regardless of the delivery mechanism, you will need to decode the received messages into Chat messages. Details of the mapping from Ably Pub/Sub messages to Chat messages are available in the [Integrations](/docs/chat/integrations) documentation.

After performing the decoding to get your chat `Message` object, you can:

1. Save it to your own database. You can index by `serial`, this is the global unique identifier for a message, and is also used to sort messages in the canonical global order.
2. If the message already exists by `serial`, it means you have received an update, delete, or reaction summary update. To check if you need to update the message, you can use the `version.serial` to compare the latest version of the message you have received with the version of the message you have in your database. Lexicographically higher means newer version.
3. If you want to store reaction summaries, always update the reactions field when receiving a reaction summary update (action `4` or `message.summary`).

<Code>
```javascript
const saveOrUpdateMessage = (message) => {
  // Check if the message already exists in your own database by `serial`

>   const existingMessage = await getMessageBySerial(message.serial);

Reviewer: Reflecting on this example - for many database engines this would be racy and may throw errors (e.g. if a message and an update happen in very quick succession). Whether or not to update would usually happen atomically at the database layer e.g. SQL's `INSERT ... ON DUPLICATE KEY UPDATE`

Author: I can see how people might be tempted to just add implementations for `getMessageBySerial`, `saveMessage` and `updateMessage` instead of actually writing a `saveOrUpdateMessage` for their own system.

That wasn't the purpose of the example when I wrote it. I meant this example to show the logic: how should you decide insert, update or discard.

I'm not sure how to fix it:

1. explain in the text that this is an example to illustrate how to compare versions/etc, not meant to be a code template
2. change it to use SQL and some assumed schema (I can mention said schema early in the article)? That might not be great for those who don't use SQL but I guess it'll be readable enough to be understood by anyone.
3. something else?

    return;
  } else if (message.version.serial === existingMessage.version.serial && message.action !== 'message.summary') {
    // if the message is the same version, and the action is not a summary, discard
    return;
  }

  // message is newer or it's a summary event => update the message
  await updateMessage(existingMessage, message);
};
```
</Code>

## Using a webhook via integration rules

Ably can forward messages to your own system via a webhook. This is the simplest to set up if you don't already have other systems in place for message ingestion. This section covers the simple HTTP endpoint webhook, but the same principles apply to other webhook integrations such as AWS Lambda, Azure Function, Google Function, and others.

Read the guide on [outbound webhooks](/docs/platform/integrations/webhooks) for more details on how to set up a webhook with Ably for the platform of your choice.

All webhook integrations allow you to use a regex filter on the channel name to control which channels the webhook should be triggered for. Use a common prefix in the name of chat rooms that you want to trigger a webhook for, and use the prefix as the filter.

> Use `channel.message` as the event type.

Author: I think that's a bit out of scope.

Maybe mention that those exist? I can imagine people might want to keep a history of when users come online so saving presence updates can be useful for that use case but I'd focus on messages for this guide? I'd say there can be another one on saving presence history if we think it's important?

---
title: "Guide: Export chat data to your own systems"
meta_description: "Learn how to export chat data from Ably Chat to your own systems."
meta_keywords: "chat, data, export, stream, storage, Ably, chat SDK, realtime messaging, dependability, cost optimisation"
---

> Ably Chat is designed to be a simple and easy to use realtime chat solution that handles any scale from 1:1 and small group chats to large livestream chats with millions of users.

Reviewer 2: ```suggestion
Ably Chat is a simple and easy to use realtime chat solution that handles any scale from 1:1 and small group chats to large livestream chats with millions of users.
```

Author: applied, thanks!

> Ably holds data for the purpose of providing realtime experiences. While Ably Chat provides flexible data retention for messages (30 days by default, up to a year on request), applications often need longer-term storage or additional control over their data.

Reviewer 2: ```suggestion
Ably holds data for the purpose of providing realtime experiences. While Ably Chat provides flexible data retention for messages (30 days by default, up to a year on request), some applications may need longer-term storage or additional control over their data.
```

More a nit pick, so feel free to ignore :)

This guide presents different ways to use Ably Chat and store chat data in your own systems, which can help you meet your data retention requirements, as well as help you build more complex use cases such as search, analytics, and more.

## Different ways to export data from Ably Chat

> We will explain each in detail, and provide code examples for each. This is an overview of the different ways to export data from Ably Chat.

Author: removed

This article covers the following options:

1. Using [outbound webhooks](/docs/platform/integrations/webhooks). [HTTP endpoint](/docs/platform/integrations/webhooks/generic), [AWS Lambda](/docs/platform/integrations/webhooks/lambda), and others.
2. Using [outbound streaming](/docs/platform/integrations/streaming). Stream to your own [Kafka](/docs/platform/integrations/streaming/kafka), [Kinesis](/docs/platform/integrations/streaming/kinesis), and others.
3. Using an [Ably queue](/docs/platform/integrations/queues).
4. Publishing via your own servers.
5. Using the Chat History endpoint.

## Decoding and storing messages

Regardless of the delivery mechanism, you will need to decode the received messages into Chat messages. Details of the mapping from Ably Pub/Sub messages to Chat messages are available in the [Integrations](/docs/chat/integrations) documentation.

> After performing the decoding and you have a chat `Message` object, you can:

Reviewer 2: ```suggestion
After performing the decoding to get your chat `Message` object, you can:
```

1. Save it to your own database. You can index by `serial`, this is the global unique identifier for a message, and is also used to sort messages in the canonical global order.
2. If the message already exists by `serial`, it means you have received an update, delete, or reaction summary update. To check if you need to update the message, you can use the `version.serial` to compare the latest version of the message you have received with the version of the message you have in your database. Lexicographically higher means newer version.

> 3. If you want to store reaction summaries, always update the reactions field when receiving a reaction summary update (action `4` or `message.summary`).

Reviewer 2: Perhaps we should add some links here to direct users to sections on message serials/versions?

Author: Not quite sure what to link to, and where to put it exactly. Maybe https://ably.com/docs/chat/rooms/messages#ordering-update-delete ?

Author: added a link in a "read more about ..." sentence. This section was changed a lot since this comment though.

1. Do you need to store only the current state of the reactions, historic snapshots of the current state, or the full history of all individual reactions?
    - If you only need the current state (latest summary), simply save the values provided in the latest message with action `message.summary`. Uniquely index by `roomName` and `serial`.
    - If you need to store historic snapshots, store all `message.summary` events for every message. Note that when a message receives many reactions in a short amount of time, summaries can be rolled up for cost and bandwidth optimisation, so not every reaction gets a summary event published.
2. Do you have a requirement to store the list of clientIds who reacted to a message, or just the totals?
    - If you only need the totals, simply use the values provided in each message with action `message.summary`.
    - If you need the list of clientIds who reacted, use the values from reaction summaries.

If you do not need to store message reactions, you can simply discard them. Never store the `reactions` (or `annotations`) field and ignore messages with action `message.summary`.

## Filtering rooms and event types <a id="filtering"/>

> Integration rules allow you to filter which Ably channels are forwarded to your own system using a regular expression on the channel name. This is a simple way to reduce the volume of messages you need to process by only receiving messages from the chat rooms you are interested in. Use a common prefix in the name of chat rooms that you want to trigger an integration rule for, and use the prefix as the filter.

Reviewer: ```suggestion
Integrations allow you to filter which Ably channels are forwarded to your own system using a regular expression on the channel name. This is a simple way to reduce the volume of messages you need to process by only receiving messages from the chat rooms you are interested in. Use a common prefix in the name of chat rooms that you want to trigger an integration rule for, and use the prefix as the filter.
```

2. Do you have a requirement to store the list of clientIds who reacted to a message, or just the totals?
    - If you only need the totals, simply use the values provided in each message with action `message.summary`.
    - If you need the list of clientIds who reacted, the values in summaries are sufficient for a reasonably large number of reactions. Over this limit, summaries will be clipped. Clipped summaries are marked with a `clipped` property set to `true`. Read more about [reaction summary clipping](/docs/chat/rooms/message-reactions#large-summaries).
    - If you predict that your reaction summaries will get clipped and you need to store the list of clientIds who reacted, consider storing individual reaction events (called _"raw reactions"_ in Chat).
    - The totals in clipped summaries represent the grand total of reactions, not just those included in the truncated clientId list.

> If you do not need to store message reactions, you can simply discard them. Never store the `reactions` (or `annotations`) field and ignore messages with action `message.summary`.

Reviewer 2: ```suggestion
If you do not need to store message reactions, you can simply discard them - i.e, by not storing the `reactions` (or `annotations`) fields and ignoring any messages with action `message.summary`.
```

Feel free to ignore/re-word, but the original sentences sounded a bit strange to me :)

Author: thanks!

---

Ably Chat is a simple and easy-to-use realtime chat solution that handles any scale from 1:1 and small group chats to large livestream chats with millions of users.

Ably holds data for the purpose of providing realtime experiences. While Ably Chat provides flexible data retention for messages (30 days by default, up to a year on request), this guide discusses options for longer-term storage or additional control over data.

This guide shows you how to export data from Ably Chat to your own systems. Before diving into the technical implementation, it's important to understand your architectural goals and what role your database will play.

## Why export chat data? <a id="why"/>

Exporting chat data from Ably Chat to your own systems can be beneficial for many use cases, for example:

- **Compliance and legal requirements**: Meet data retention policies, maintain audit trails for support conversations, or fulfill regulatory requirements.
- **Analytics and business intelligence**: Build dashboards, train ML models, analyze customer sentiment, or track support quality metrics.
- **Enhanced functionality**: Implement features that need the chat history, such as search.

> - **Data sovereignty**: Maintain your own database as the canonical source of truth.

Author: changed

- **Database schema**: Design your schema to allow you to easily build the features you need, keeping in mind scale and reliability.
  - **Version history requirements**: Decide whether you need to store all versions of messages or just the latest version (see [Decision 1](#decision-1) below). If you need all versions, you'll need to design your schema accordingly. If you want to retrieve messages as shown in a chat window (latest version of each message), consider duplicating the latest message data in a separate table to avoid filtering at query time.
  - **Indexing for search**: consider indexing only the latest version of each message, or have a way to indicate which search results represent current versions.
  - **Concurrent writes**: New messages, updates, and deletes will arrive concurrently, so your database system must handle this. Depending on your database, consider reducing roundtrips, managing locks, and handling race conditions.
- **Scale and reliability trade-offs**: Depending on the scale of your application, you need to consider how you will scale up and down the parts that handle the ingestion of messages from Ably.

> - **Data latency and consistency**: If you publish directly to Ably (recommended), there will be a small delay between a message being published and it arriving in your database via integrations. Think of how to mitigate if needed for situations where the most up-to-date data is required.

Author: @splindsay-92  I've rephrased, what do you think of it now? I just imply that by default publishing is via ably (which is likely the case) and offer publish via your own as a solution to data-in-your-db latency if needed.

Removed the `(recommended)`, it was a bit out of place.
